doc/dev/cephadm/design/storage_devices_and_osds.rst

```diff
@@ -0,0 +1,317 @@
+==============================================
+Storage Devices and OSDs Management Workflows
+==============================================
+The cluster storage devices are the physical storage devices installed in each of the cluster’s hosts. We need to execute different operations over them and also to retrieve information about physical features and working behavior.
+The basic use cases we have in this area are:
+
+- `1. Retrieve device information. Inventory`_
+- `2. Add OSDs`_
+- `3. Remove OSDs`_
+- `4. Replace OSDs`_
+
+1. Retrieve device information. Inventory
+=========================================
+We must be able to review what is the current state and condition of the cluster storage devices. We need the identification and features detail (including ident/fault led on/off capable) and if the device is used or not as an OSD/DB/WAL device.
+
+The information required for each device will be at least:
+::
+
+    Hostname   Path      Type  Serial    Size   Health  Ident  Fault  Available
+
+.. Note: A more optional extended view with other information fields could be also useful.
+
+In order to know what is the current condition of the device, we need to know what is the amount of storage used, the percentage of free space, the average number of IOPS and the fault led state.
+This information should be provided probably by the Ceph orchestrator which is the component where we have access to this kind of information.
+
+Another important question around retrieving device information is “efficiency”. The information about devices can be critical in components like the Orchestrator or the Dashboard, because this information usually is used to take decisions.
+When we talk about efficiency we need to be sure that all the points are covered:
+
+#. Get the complete information for each device in the most fast way.
+#. All the information about all the devices in one host is accessible always immediately.
+#. The information is constantly updated in each host. A device failure or the addition of a new device must be detected in the smallest possible timeframe
+#. Scalability. To work with thousands of devices in hundreds of hosts shouldn't be a problem.
+
+A. Current workflow:
+--------------------
+**CLI**:
+    Operations:
+
+.. prompt:: bash #
+
+  ceph orch device ls
+  ceph orch device ls json ( to get all the fields for each device )
+
+    Problems in current implementation:
+        * Does not scale.
+
+**GUI**:
+    Operations:
+        * cluster.Inventory section:
+            The cluster.Inventory section presents a basic list of the devices in the cluster. It is a fixed list with only a few fields. Only the “ident light on” operation is possible although we do not know if it is possible or not until the operation is launched.
+
+    Problems in current implementation:
+        * Does not scale (depends of the orchestrator)
+        * Rigid user experience
```

Comment:
have a look at  https://docs.ceph.com/en/latest/api/mon_command_api/#orch-device-ls

We already have:

* `--wide`
* `--hostname`

is that ok for you?

Reply:
@sebastian-philipp : Yes, it is nice to have these possibilities in the "`device ls`" command. We have some kind of filtering using the `hostname` an more information using the `wide` format or the json output.

But in this section what we are doing **is pointing problems of scalability and UX around the list of devices** in the cluster.
This document was written 4 months ago ... and some things have changed a slightly:
- 4 months ago, to get information from the different devices in the cluster was not very scalable.... we have improve in this part thanks to the cephadm exporter.
- Rigid User experience, afaik nothing has changed in the user interface to make possible the final user to select a set of fields from the "bag" of fields provided for each device using "ceph orch ls --format json"

Reply:
@jmolmo shall we merge this then?